src/gamescope/wayland/metrics/dbus.rs

```diff
@@ -1,39 +1,82 @@
-use std::time::SystemTime;
-
-use gamescope_wayland_client::mangoapp::{
-    message::MangoAppMsgV1,
-    queue::{MangoAppMsgQueue, QueueError},
+use std::{
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    thread::JoinHandle,
+    time::{Duration, SystemTime},
 };
+
+use gamescope_wayland_client::mangoapp::message::MangoAppMsgV1;
 use thiserror::Error;
+use tokio::{
+    runtime::Runtime,
+    time::{error::Elapsed, timeout},
+};
 use zbus::{dbus_interface, fdo};
 
+use crate::gamescope::wayland::metrics::mango_app_reader::MangoAppMsgQueueReader;
+
 #[derive(Error, Debug)]
 pub enum MetricsError {
     #[error("failed to read metrics message from gamescope: {0}")]
-    QueueError(#[from] QueueError),
+    QueueError(#[from] tokio::sync::watch::error::RecvError),
+    #[error("timedout trying to read gamescope queue message: {0}")]
+    QueueTimeout(#[from] Elapsed),
 }
 
 impl From<MetricsError> for fdo::Error {
     fn from(value: MetricsError) -> Self {
         match value {
             MetricsError::QueueError(queue_error) => fdo::Error::Failed(queue_error.to_string()),
+            MetricsError::QueueTimeout(queue_error) => fdo::Error::Failed(queue_error.to_string()),
         }
     }
 }
 
 pub struct DBusInterface {
-    queue: MangoAppMsgQueue,
+    handle: Option<JoinHandle<()>>,
+    cancel_flag: Arc<AtomicBool>,
     state: Option<MangoAppMsgV1>,
     last_update: SystemTime,
+    request_tx: std::sync::mpsc::Sender<()>,
+    response_rx: tokio::sync::watch::Receiver<Option<MangoAppMsgV1>>,
+}
+
+impl Drop for DBusInterface {
+    fn drop(&mut self) {
+        self.cancel_flag.store(true, Ordering::Relaxed);
+        let _ = self.request_tx.send(());
+
+        if let Some(handle) = self.handle.take() {
+            let _ = handle.join();
+        }
+    }
 }
 
 impl DBusInterface {
     pub fn new() -> Self {
-        let queue = MangoAppMsgQueue::new();
+        let (response_tx, response_rx) = tokio::sync::watch::channel::<Option<MangoAppMsgV1>>(None);
+        let (request_tx, request_rx) = std::sync::mpsc::channel::<()>();
+        let cancel_flag = Arc::new(AtomicBool::new(false));
+        let cancel_flag_thread = cancel_flag.clone();
+
+        let handle = std::thread::spawn(move || {
+            let rt = Runtime::new().expect("Failed to create runtime");
+
+            rt.block_on(async move {
+                let reader = MangoAppMsgQueueReader::new(request_rx, response_tx);
+                reader.run(cancel_flag_thread).await;
+            });
+        });
```

Comment:
Doing this will spawn a new thread and create an entirely new and separate Tokio runtime to run this single task. We should instead use `tokio::task::spawn_blocking()` to spawn a separate blocking thread to run this in.

```suggestion
        tokio::task::spawn_blocking(move || {
            let reader = MangoAppMsgQueueReader::new(request_rx, response_tx);
            reader.run();
        });
```